test/integration/puppetlabs/services/master/environment_modules_int_test.clj

```diff
@@ -0,0 +1,392 @@
+           ;;    the 'test' environment was not flushed.
+           (purge-env-dir)
+           (purge-all-env-caches)
+           ;; add version that has multiple module versions
+           (let [prod-file (testutils/write-pp-file "class oneprod {}"
+                                                    "somemod"
+                                                    "prod"
+                                                    "production"
+                                                    "1.0.0")
+                 test-file (testutils/write-pp-file "class onetest {}"
+                                                    "somemod"
+                                                    "test"
+                                                    "test"
+                                                    "2.0.0")
+                 production-response-initial (get-env-modules "production")
+                 test-response-initial (get-env-modules "test")]
+             (is (= 200 (:status production-response-initial))
+                 (str
+                  "unexpected status code for initial production response"
+                  "response: "
+                  (ks/pprint-to-string production-response-initial)))
+             (is (= {"modules" [{"name" "somemod", "version" "1.0.0"}]
+                     "name" "production"}
+                    (response->module-info-map production-response-initial))
+                 "unexpected body for production response")
+             (is (= 200 (:status test-response-initial))
+                 (str
+                  "unexpected status code for initial test response"
+                  "response: "
+                  (ks/pprint-to-string test-response-initial)))
+             (is (= {"modules"[{"name" "somemod", "version" "2.0.0"}]
+                     "name" "test"}
+                    (response->module-info-map test-response-initial))
+                 "unexpected body for test response")
+
+             (testutils/write-pp-file "class oneflushprod {}"
+                                      "somemod"
+                                      "prod"
+                                      "production"
+                                      "1.0.0")
+             (testutils/write-pp-file "class oneflushtest {}"
+                                      "somemod"
+                                      "test"
+                                      "test"
+                                      "2.0.0")
+
+             (let [production-response-before-flush (get-env-modules
+                                                     "production")
+                   test-response-before-flush (get-env-modules
+                                               "test")]
+               (is (= 304 (:status production-response-before-flush))
+                   (str
+                    "unexpected status code for prod response after code change "
+                    "but before flush"))
```

Comment:
@briancain I wouldn't have expected the caching to work with the environment_modules endpoint as this PR currently stands.

For the environment_classes endpoint, the middleware functions hash the response body and include an http etag, like [here](https://github.com/puppetlabs/puppetserver/blob/2.7.2/src/clj/puppetlabs/services/master/master_core.clj#L292-L301).  Since you don't have any of that in the environment_modules middleware functions, I wouldn't expect the endpoint to return a 304 when the module content doesn't change from one query to the next.

I don't necessarily think you need to bother cache-enabling the environment_modules endpoint at this point.  Assuming you aren't planning on having the analytics service query for this info super frequently, the performance without caching may be fine.  Enumerating module info should be a lot less heavy than what the environment_classes endpoint has to do to parse manifest info anyway and the caching layer adds a fair bit of complexity.

Setting aside the caching, I think you could thin these extra tests out considerably.  I'd be fine with just a single test which validates that a 200 OK is returned with a valid json payload containing modules and version info for a couple of different environments.